ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -966,46 +1098,88 @@ private void handleMetadataBlob(BlobData blobData) throws IOException, MessageFo
       } catch (IllegalArgumentException e) {
         onInvalidRange(e);
         rangeResolutionFailure = true;
+        blobContentAvailableToProcess = true;
       }
       if (!rangeResolutionFailure) {
         if (options.getChunkIdsOnly) {
           chunkIdIterator = null;
           numChunksTotal = 0;
           dataChunks = null;
+          blobContentAvailableToProcess = true;
         } else {
           chunkIdIterator = keys.listIterator();
           numChunksTotal = keys.size();
           dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
           for (int i = 0; i < dataChunks.length; i++) {
             dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
           }
+          // if blob is encrypted, then decryption is required only in case of GetBlobInfo and GetBlobAll (since user-metadata
+          // is expected to be encrypted). Incase of GetBlob, Metadata blob does not need any decryption even if BlobProperties says so
+          if (getOperationFlag() != MessageFormatFlags.Blob && blobInfo.getBlobProperties().isEncrypted()) {
```

Comment:
Wouldn't blobInfo always be non-null if the OperationFlag is MessageFormatFlags.ALL?